---
title: "MIA_main_markdown"
author: "Jin Wang"
date: "2020/9/22"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r libraries, message = FALSE}
source("~/Documents/Description_study/src_figures_NC/MIA_functions.R")
```

```{r,fig.height = 4, fig.width = 9}
#####################----------------Figure 1: GSEA for differential expressed genes-----------------##########################
###############################################################################################################################
##KEGG
load(file = "~/Documents/Description_study/data/MIA_GSEA_KEGG.Rdata")
load(file = "~/Documents/Description_study/data/KEGG_pathway_category.Rdata")
cate <- "KEGG Pathways"
highlight <- c("Mucin type O-glycan biosynthesis","Cytokine-cytokine receptor interaction") #"ABC transporters",
kegg.terms <- GSEA.KEGG %>% 
  mutate(logP = ifelse(NES >= 0, -log10(p.adjust), log10(p.adjust))) %>%
  mutate(ratio = (str_count(core_enrichment,"/")+1)/setSize) %>%
  dplyr::filter(p.adjust <= 0.05 & ratio > 0.25) %>% 
  mutate(text_col = ifelse(Description %in% highlight, "red", "black"))
kegg.terms$category <- sapply(kegg.terms$ID, function(t){
  cate <- unlist(plyr::compact(lapply(names(pathway_id), function(x) if(t %in% pathway_id[[x]]) return(x))))
  return(gsub(": [a-z]+", "",cate))
})
kegg.terms.order <- kegg.terms %>% 
  subset(select = c("Description", "NES","category","significance", "text_col")) %>%
  dplyr::filter(!category %in% c(" Development and regeneration", " Endocrine and metabolic disease", " Infectious disease", " Transport and catabolism")) %>%
  dplyr::arrange(category, desc(NES)) %>%
  top_n(15, NES) %>%
  mutate(Description = ifelse(Description == "Signaling pathways regulating pluripotency of stem cells", "Signaling of pluripotency of stem cells", as.character(Description)))
###order categories
terms.order <- rev(c(" Signaling molecules and interaction", " Signal transduction", " Membrane transport", " Glycan biosynthesis and metabolism", " Cellular community - eukaryotes", " Cancer"))
kegg.terms.order.cate <- do.call("rbind",lapply(terms.order, function(t) subset(kegg.terms.order, category == t)))
kegg.terms.order.cate$Description <- factor(kegg.terms.order.cate$Description, levels = kegg.terms.order.cate$Description)  
###barplot
ggplot(kegg.terms.order.cate,aes(x= Description, y = NES, fill = category))+
  geom_bar(stat = "identity",alpha = 0.8)+
  theme_bw()+
  # scale_y_continuous(breaks = brks, labels = lbls)+
  scale_fill_manual(values = brewer.pal(n = 10, name = "Set3"), name = "Pathway Category")+
  labs(y = "NES", x = NULL)+
  annotate("text",label=kegg.terms.order$significance,
           y=ceiling(max(kegg.terms.order$NES)),
           x=1:length(unique(kegg.terms.order$Description)), size = 5)+
  theme(legend.position = "right",
        axis.text.x=element_text(size = 12,face = "bold"),
        axis.text.y = element_text(size = 12,face = "bold",colour = kegg.terms.order.cate$text_col),
        axis.title.x = element_text(size = 12,face = "bold"),
        axis.title.y = element_text(size = 12, face = "bold"),
        axis.ticks.x.bottom = element_blank()
  )+
  coord_flip()
```


```{r, fig.height = 5, fig.width = 10}
#####################----------------Figure 1: MUC alterations-----------------##########################
######################################################################################################################################
##read in maf files and write into oncomatrix
###specify genes
genelist.filter <- read.table("~/Documents/Description_study/data/MUC_genes.txt")
genelist.filter <- as.character(genelist.filter[,1])
###read in MIA mutations
oncomatrix.list <- get(load("~/Documents/Description_study/vcf2maf_data/lung_merged_59_oncomatrix.Rdata"))
oncomatrix.output <- oncomatrix.list$oncoMatrix
colnames(oncomatrix.output) <- paste("Pat",pat.info[match(colnames(oncomatrix.output), pat.info$value),'patientID'],sep = "")
ss <- colnames(oncomatrix.output)
sub.oncomatrix <- oncomatrix.output[match(genelist.filter,rownames(oncomatrix.output)),]
sub.oncomatrix[is.na(sub.oncomatrix)] <- ""
rownames(sub.oncomatrix) <- genelist.filter
sub.oncomatrix.new <- apply(sub.oncomatrix,c(1,2),function(x){
  x <- ifelse(grepl("Missense",x),"MISSENSE",
              ifelse(grepl("In_Frame",x), "INFRAME",
                     ifelse(grepl("Frame_Shift|Nonsense|Splice",x),"TRUNC",
                            ifelse(grepl("Multi_Hit",x),"MULTI_HIT",""))))
  return(x)
})

###read in MIA expression data
n.mean <- apply(t.TPM[,n.sample],1,mean)
n.sd <- apply(t.TPM[,n.sample],1,sd)
expr.norm <- (t.TPM[,t.sample]-n.mean)/n.sd
colnames(expr.norm) <- paste("Pat",pat.info[match(colnames(expr.norm), gsub("-",".",pat.info$value)),'patientID'],sep = "")
expr.norm <- expr.norm[,ss] ###remove an outlier
expr.norm <- ifelse(expr.norm >= 2,"UP",ifelse(expr.norm <= -2,"DOWN",""))
sub.expr <- expr.norm[genelist.filter,]
###combine mutation and expression data
datArray <- abind(sub.oncomatrix.new,sub.expr,along = 3)
dat <- apply(datArray,1:2,function(x) 
  if(x[1] == "" & x[2] == "") {return ("")}
  else if(x[1] == "" & x[2] != ""){paste(x[2],";",sep = "")}
  else if(x[1] != "" & x[2] == ""){paste(x[1],";",sep = "")}
  else {paste(x[1],x[2],sep = ";")})
###oncoprint
alter_fun = list(
  background = function(x, y, w, h) {
    grid.rect(x, y, w-unit(0.5, "mm"), h-unit(0.5, "mm"), gp = gpar(fill = "#CCCCCC", col = NA))
  },
  DOWN = function(x, y, w, h) {
    grid.rect(x, y, w-unit(0.5, "mm"), h-unit(0.5, "mm"), gp = gpar(fill = NA, col = "cornflowerblue"))
  },
  UP = function(x, y, w, h) {
    grid.rect(x, y, w-unit(0.5, "mm"), h-unit(0.5, "mm"), gp = gpar(fill = NA, col = "lightcoral"))
  },
  MISSENSE = function(x, y, w, h) {
    grid.rect(x, y, w-unit(0.5, "mm"), h*0.33, gp = gpar(fill = "#33A02C", col = NA))
  },
  MULTI_HIT = function(x, y, w, h) {
    grid.rect(x, y, w-unit(0.5, "mm"), h*0.33, gp = gpar(fill = "black", col = NA))
  },
  INFRAME = function(x, y, w, h) {
    grid.rect(x, y, w-unit(0.5, "mm"), h*0.33, gp = gpar(fill = "#A6CEE3", col = NA))
  },
  TRUNC = function(x, y, w, h) {
    grid.rect(x, y, w-unit(0.5, "mm"), h*0.33, gp = gpar(fill = "gray70", col = NA))
  }
)
##row barplot
##gene order
alter.perc <- apply(dat, 1, function(x)length(x[x != ""]))
ge.order <- names(sort(alter.perc[alter.perc != 0],decreasing = TRUE))
##sample order
alter.perc <- apply(dat, 2, function(x)length(x[x != ""]))
sa.order <- names(sort(alter.perc,decreasing = TRUE))
##annotation
col_fun = structure(names = c("DOWN","UP",'MISSENSE','INFRAME', 'TRUNC', 'MULTI_HIT'),
                    c("cornflowerblue","lightcoral","#33A02C","#A6CEE3","slategray4","black"))
###oncoprint
ht_list <- oncoPrint(dat[ge.order,], get_type = function(x) strsplit(x, ";")[[1]],
                     alter_fun = alter_fun, col = col_fun, row_order = ge.order,width = unit(12, "cm"),
                     row_names_gp = gpar(fontsize = 14,fontface = "bold"),pct_gp = gpar(fontsize = 14,fontface = "bold"),
                     remove_empty_columns = FALSE,
                     top_annotation = NULL, 
                     right_annotation = rowAnnotation(rbar = anno_oncoprint_barplot(type = names(col_fun),which = "row",bar_width = 0.8, axis = TRUE,width = unit(1.5, "cm"))),
                     heatmap_legend_param = list(title = "Alterations", at = names(col_fun), labels = names(col_fun)))
draw(ht_list)
```

```{r}
#####################----------------Figure 2: immune infiltration from TIMER in Chinese MIA-----------------#################
###############################################################################################################################
Infil.est <- read.csv(file = '~/Documents/BCR_Project_Xiaoqi/20171024-RNA-Seq data/score_matrix.csv',header = TRUE)
Infil.est <- subset(Infil.est, !(sampleID %in% c("FZ.25","FZ.26",'FZ.31','FZ.32')))
CompareInfiltration(Infil.est, 6, pat.diagnosis,pair = TRUE,title = "MIA", tool = "TIMER", xsize = 12)
```



```{r}
#####################----------------Figure 2: immune infiltration from TIMER for TCGA different stage samples-------------##########################
######################################################################################################################################
cc <- "LUAD"
##clinical data
clin <- read.table(file = paste("~/Documents/BCR_Project_Xiaoqi/TRUST3_BCR_aliyun/TCGA_Lung/",cc,".merged_only_clinical_clin_format.txt",sep = ""),sep = "\t",header = TRUE,fill = TRUE, row.names = 1)
stage <- sub("(a|b|c)$","",t(clin["patient.stage_event.pathologic_stage",])[,1])
names(stage) <- toupper(t(clin["patient.bcr_patient_barcode",])[,1])
##only extract tumor samples
cc.timer <- read.table(file = paste("~/Documents/CIDC/VALIDATION/TCGA_Infiltration/normalized_RSEM_Timer/",cc,"_TPM_normalized_timer_output.tsv",sep = ""),sep = "\t",header = TRUE,row.names = 1)
if(grepl("\\.",rownames(cc.timer))[1]){
  rownames(cc.timer) <- gsub("\\.","-",rownames(cc.timer))
}
if(!grepl("\\.",rownames(cc.timer))[1]){
  rownames(cc.timer) <- rownames(cc.timer)
}
cc.timer <- data.frame(sampleID = rownames(cc.timer),cc.timer)
cc.timer.n <- subset(cc.timer,substr(rownames(cc.timer),14,14) != 0)
cc.timer.t <- subset(cc.timer,substr(rownames(cc.timer),14,14) == 0)
stage.gr <- stage[substr(rownames(cc.timer.t),1,12)]
cc.timer.new <- rbind.data.frame(cc.timer.t[which(!is.na(stage.gr)),], cc.timer.n)
gr <- c(na.omit(stage.gr),rep("Normal",dim(cc.timer.n)[1]))
##plot
Infil.est <- cc.timer.new
n <- 6
diagnosis <- gr
pair <- FALSE
title <- "TCGA LUAD"
xsize <- 12
mycomparisons <- list(c("Normal", "stage i"),c("Normal","stage ii"), c("Normal", "stage iii"), 
                      c("stage i", "stage ii"), c("stage i", "stage iii"),c("stage i", "stage iv"))
infil.melt <- melt(Infil.est,id.vars = "sampleID")
colnames(infil.melt) <- c("ID","Immune.cell","Estimation")
infil.melt$Group <- c(rep(diagnosis,n))
infil.melt.new <- subset(infil.melt, Immune.cell %in% c("B_cell", "T_cell.CD4", "T_cell.CD8"))
ggplot(infil.melt.new,aes(x = Group, y=Estimation, fill = Group))+
  geom_boxplot(outlier.size = 0.5)+
  # ggtitle(title)+ 
  scale_fill_manual(values = c("#377EB8",brewer.pal(n = 9, name = "YlOrRd")[c(4,6,8,9)]))+
  #annotate("text",x = 1:n, y = as.numeric(apply(Infil.est[2:(n+1)],2,max))+0.1, label = p.star)+
  theme_bw()+
  facet_wrap(~Immune.cell) + 
  ylab("Immune cell Abundance")+
  xlab("Immune Cell")+
  ylim(c(0,0.3))+
  stat_compare_means(aes(group = Immune.cell), label = "p.signif", label.y = seq(0.18, 0.28, by = 0.02), method = "wilcox.test",
                     comparisons = mycomparisons,color = "red", size = 4,
                     symnum.args = list(cutpoints = c(0, 0.001, 0.01, 0.05, 1), symbols = c( "***", "**", "*", ""))
  )+
  theme(plot.title = element_text(hjust=0.5,vjust = 0.5, margin = margin(l=100,r=50,t=10,b=10),face = "bold", colour = "black"),
        axis.text.x=element_text(angle=30,size=xsize,face = "bold",hjust=1),
        axis.text.y=element_text(size=xsize,face = "bold",hjust=1),
        axis.title.x = element_blank(),
        axis.title.y = element_text(size = 12,face = "bold"),
        legend.text = element_text(size = 12),
        strip.text = element_text(size = 12, face = "bold"),
        legend.position = "bottom") 
```


```{r,fig.height=3, fig.width=3}
#####################----------------Figure 2: TCR metrics Chinese MIA-------------##########################
####################################################################################################################################
#################compare unique cdr3s size
cdr3.size <- aggregate(cdr3aa ~ filename+ID+Tissue, TCRdata, function(x) length(unique(x))) 
cdr3.size$ID <- as.numeric(cdr3.size$ID)
all.cdr3.size <- arrange(cdr3.size,ID)
t.cdr3 <- all.cdr3.size[match(tumor,all.cdr3.size$ID),]
n.cdr3 <- all.cdr3.size[match(normal,all.cdr3.size$ID),]
all.cdr3 <- rbind(t.cdr3,n.cdr3) %>% 
  mutate(Tissue = as.character(sapply(ID,function(x) if(x %in% normal) {return("Normal")} else{return("Tumor")})),
         Tissue = factor(Tissue,levels = c("Tumor","Normal")))
value.cdr3 <- signif(wilcox.test(t.cdr3$cdr3aa,n.cdr3$cdr3aa,paired = TRUE)$p.value,2)
value.cdr3
text <- as.character(cut(value.cdr3, right = FALSE, breaks = c(0, 0.001, 0.01, 0.05,1), labels = c("***","**","*"," ")))
gc <- ggviolin(all.cdr3, "Tissue", "cdr3aa", fill = "Tissue",
               palette = c("#E41A1C","#377EB8"), trim = TRUE,
               add = "boxplot", add.params = list(fill = "white"))+
  annotate("text",label = text,x = 1.55, y = max(all.cdr3$cdr3aa)*0.8,size = 7, color = "red")+
  #annotate("text",label = paste("P = ",value.cdr3,sep = ''),x = 1.55, y = max(all.cdr3$CDR3_aa)*0.8)+
  theme_bw()+
  ylab("#Unique TCR CDR3")+
  xlab("")+
  theme(legend.position = "none",
        axis.text.x=element_text(angle=0,size=12,face = "bold",hjust=0.5),
        axis.text.y = element_text(size = 12,face = "bold"),
        axis.title.x = element_blank(),
        axis.title.y = element_text(size = 12,face = "bold"))
print(gc)
```


```{r, fig.height=3, fig.width=3}
#####################----------------Figure 2: BCR metrics for Chinese MIA-------------##########################
####################################################################################################################################
##calculate unique cdr3s size
cdr3.size <- aggregate(CDR3_aa ~ filename+ID+Tissue, BCRdata, function(x) length(unique(x))) 
cdr3.size$ID <- as.numeric(cdr3.size$ID)
all.cdr3.size <- arrange(cdr3.size,ID)
t.cdr3 <- all.cdr3.size[match(tumor,all.cdr3.size$ID),]
n.cdr3 <- all.cdr3.size[match(normal,all.cdr3.size$ID),]
all.cdr3 <- rbind(t.cdr3,n.cdr3) %>% 
  mutate(Tissue = as.character(sapply(ID,function(x) if(x %in% normal) {return("Normal")} else{return("Tumor")})),
         Tissue = factor(Tissue,levels = c("Tumor","Normal")))
value.cdr3 <- signif(wilcox.test(t.cdr3$CDR3_aa,n.cdr3$CDR3_aa,paired = TRUE)$p.value,2)
text <- as.character(cut(value.cdr3, right = FALSE, breaks = c(0, 0.001, 0.01, 0.05,1), labels = c("***","**","*"," ")))
gc <- ggviolin(all.cdr3, "Tissue", "CDR3_aa", fill = "Tissue",
               palette = c("#E41A1C","#377EB8"), trim = TRUE,
               add = "boxplot", add.params = list(fill = "white"))+
  annotate("text",label = text,x = 1.55, y = max(all.cdr3$CDR3_aa)*0.8,size = 7, color = "red")+
  theme_bw()+
  ylab("#Unique BCR CDR3")+
  xlab("")+
  theme(legend.position = "none",
        axis.text.x=element_text(angle=0,size=12,face = "bold",hjust=0.5),
        axis.text.y = element_text(size = 12,face = "bold"),
        axis.title.x = element_blank(),
        axis.title.y = element_text(size = 12,face = "bold"))
print(gc)
```

```{r,fig.height=3, fig.width=3}
### compare cluster number between tumor and normal
bcr.cluster.num <- data.frame(do.call("rbind",lapply(bcr_clusters, length)))
colnames(bcr.cluster.num) <- "cluster_num"
bcr.cluster.num$ID <- as.numeric(gsub("FZ-(.*)Aligned\\.sortedByCoord\\.out\\.bam","\\1",rownames(bcr.cluster.num)))
t.bcr.cluster.num <- bcr.cluster.num[match(tumor,bcr.cluster.num$ID),]
n.bcr.cluster.num <- bcr.cluster.num[match(normal,bcr.cluster.num$ID),]
all.bcr.cluster.num <- rbind(n.bcr.cluster.num,t.bcr.cluster.num) %>% 
  mutate(Tissue = as.character(sapply(ID,function(x) if(x %in% normal) {return("Normal")} else{return("Tumor")})),
         Tissue = factor(Tissue,levels = c("Tumor","Normal")))
value.bcr.cluster.num <- signif(wilcox.test(t.bcr.cluster.num$cluster_num,n.bcr.cluster.num$cluster_num,paired = TRUE)$p.value,2)
text <- as.character(cut(value.bcr.cluster.num, right = FALSE, breaks = c(0, 0.001, 0.01, 0.05,1), labels = c("***","**","*"," ")))
value.bcr.cluster.num
all.bcr.cluster.num$Tissue <- factor(all.bcr.cluster.num$Tissue, levels = c("Tumor","Normal"))
gcn <- ggviolin(all.bcr.cluster.num, "Tissue", "cluster_num", fill = "Tissue",
               palette = c("#E41A1C","#377EB8"),trim = TRUE,
               add = "boxplot", add.params = list(fill = "white"))+
  annotate("text",label = text,x = 1.55, y = max(all.bcr.cluster.num$cluster_num),size = 7, color = "red")+
  theme_bw()+
  ylim(c(0,max(all.bcr.cluster.num$cluster_num)))+
  xlab("")+
  ylab("#BCR Clusters")+
  theme(legend.position = "none",
        axis.text.x=element_text(angle=0,size=12,face = "bold",hjust=0.5),
        axis.text.y = element_text(size = 12,face = "bold"),
        axis.title.x = element_blank(),
        axis.title.y = element_text(size = 12,face = "bold"))
print(gcn)
```


```{r,fig.height=3, fig.width=3}
########----------------------------------compare BCR CPK between tumor and normal--------------------------------------#####
load(file = "~/Documents/Description_study/data/TRUST3_MIA_BCRdata_118.Rdata")
cpk <- aggregate(CDR3_aa ~ filename+ID+Tissue+est_lib_size, BCRdata, function(x) length(unique(x))) %>%
  mutate(CPK = signif(CDR3_aa/(est_lib_size/1000),4))
cpk.all <- arrange(cpk,ID)
t.cpk <- cpk.all[match(tumor,cpk.all$ID),]
n.cpk <- cpk.all[match(normal,cpk.all$ID),]
all.cpk <- rbind(t.cpk,n.cpk) %>% 
  mutate(Tissue = factor(Tissue,levels = c("Tumor","Normal")))
value.cpk <- signif(wilcox.test(t.cpk$CPK,n.cpk$CPK,paired = TRUE)$p.value,2)
value.cpk
text <- as.character(cut(value.cpk, right = TRUE, breaks = c(0, 0.001, 0.01, 0.05,1), labels = c("***","**","*"," ")))
gk <- ggviolin(all.cpk, "Tissue", "CPK", fill = "Tissue",
               palette = c("#E41A1C","#377EB8"), trim = TRUE,
               add = "boxplot", add.params = list(fill = "white"))+
  annotate("text",label = text,x = 1.55, y = max(all.cpk$CPK)*0.9,size = 7, color = "red")+
  #annotate("text",label = paste("P = ",value.cdr3,sep = ''),x = 1.55, y = max(all.cdr3$CDR3_aa)*0.8)+
  theme_bw()+
  ylab("BCR Clontypes per kilo reads")+
  xlab("")+
  theme(legend.position = "none",
        axis.text.x=element_text(angle=0,size=12,face = "bold",hjust=0.5),
        axis.text.y = element_text(size = 12,face = "bold"),
        axis.title.x = element_blank(),
        axis.title.y = element_text(size = 12,face = "bold"))
print(gk)
```


```{r,fig.height=3, fig.width=3}
##########compare clonality between tumor and normal
clonality.all <- data.frame(clonality.all) %>% mutate(ID = as.numeric(gsub("FZ-(.*)Aligned\\.sortedByCoord\\.out\\.bam","\\1",filename)), 
                                                      clonality = as.numeric(as.character(clonality)))
t.clonality <- clonality.all[match(tumor,clonality.all$ID),]
n.clonality <- clonality.all[match(normal,clonality.all$ID),]
all.clonality <- rbind(n.clonality,t.clonality) %>% 
  mutate(Tissue = as.character(sapply(ID,function(x) if(x %in% normal) {return("Normal")} else{return("Tumor")})),
         Tissue = factor(Tissue,levels = c("Tumor","Normal")))
value.clonality <- signif(wilcox.test(t.clonality$clonality,n.clonality$clonality,paired = TRUE)$p.value,2)
text <- as.character(cut(value.clonality, right = FALSE, breaks = c(0, 0.001, 0.01, 0.05,1), labels = c("***","**","*"," ")))
value.clonality
all.clonality$Tissue <- factor(all.clonality$Tissue, levels = c("Tumor","Normal"))
# pdf(file = "~/Documents/Description_study/figures_v2/figure4_BCR_metrics_BCR_clonality.pdf",width = 2.3, height = 2.5, useDingbats = FALSE)
gn <- ggviolin(all.clonality, "Tissue", "clonality", fill = "Tissue",
               palette = c("#E41A1C","#377EB8"),trim = TRUE,
               add = "boxplot", add.params = list(fill = "white"))+
  annotate("text",label = text,x = 1.55, y = 0.10*0.8,size = 7, color = "red")+
  theme_bw()+
  ylim(c(0,0.10))+
  xlab("")+
  ylab("BCR Clonality")+
  theme(legend.position = "none",
        axis.text.x=element_text(angle=0,size=12,face = "bold",hjust=0.5),
        axis.text.y = element_text(size = 12,face = "bold"),
        axis.title.x = element_blank(),
        axis.title.y = element_text(size = 12,face = "bold"))
print(gn)
```

```{r,fig.height=3, fig.width=3}
###calculate somatic hypermutation rate
all.SHM <- cbind.data.frame(ID = as.numeric(gsub("FZ-([0-9]+).*","\\1",names(SHMRatio))), SHM.rate = unlist(SHMRatio))
t.SHM <- all.SHM[match(tumor,all.SHM$ID),]
n.SHM <- all.SHM[match(normal,all.SHM$ID),]
sub.SHM <- rbind.data.frame(t.SHM,n.SHM) %>% 
  mutate(Tissue = as.character(sapply(ID,function(x) if(x %in% normal) {return("Normal")} else{return("Tumor")})),
         Tissue = factor(Tissue,levels = c("Tumor","Normal")))
value.SHM <- signif(wilcox.test(t.SHM$SHM.rate,n.SHM$SHM.rate,paired = TRUE)$p.value,2)
text <- as.character(cut(value.SHM, right = FALSE, breaks = c(0, 0.001, 0.01, 0.05,1), labels = c("***","**","*"," ")))
value.SHM
##violin plot
gs <- ggviolin(sub.SHM, "Tissue", "SHM.rate", fill = "Tissue",
               palette = c("#E41A1C","#377EB8"),trim = TRUE,
               add = "boxplot", add.params = list(fill = "white"))+
  annotate("text",label = text,x = 1.55, y = max(na.omit(sub.SHM$SHM.rate))*0.8,size = 7, color = "red")+
  theme_bw()+
  xlab("")+
  ylab("SHM rate")+
  theme(legend.position = "none",
        axis.text.x=element_text(angle=0,size=12,face = "bold",hjust=0.5),
        axis.text.y = element_text(size = 12,face = "bold"),
        axis.title.x = element_blank(),
        axis.title.y = element_text(size = 12,face = "bold"))
print(gs)
```



```{r, fig.height=3.5, fig.width=6}
##############-------------------Figure 2: compare CD4+ T cell activation marker signal in boxplot----------------------##########
#######################################################################################################################
CD4.genes <- c("CD69","CD40LG","IL2RA","SELL")
load(file = "~/Documents/Description_study/data/CD4_markers_expr.Rdata")
CD4.data.melt <- reshape2::melt(CD4.data) 
colnames(CD4.data.melt) <- c("Gene","ID","Expression")
CD4.data.melt <- CD4.data.melt %>% mutate(tissue = ifelse(ID %in% t.sample, "Tumor","Normal"))
CD4.data.melt$Expression <- as.numeric(as.character(CD4.data.melt$Expression))
all.p <- NULL
for (i in CD4.genes){
  df <- subset(CD4.data.melt,Gene == i)
  test <- t.test(Expression~tissue,data = df) #method = "paired"
  p <- signif(test$p.value,4)
  all.p <- c(all.p,p)
}
p.star <- cut(all.p,breaks = c(-Inf, 0.001, 0.01, 0.05, Inf), label=c("***", "**", "*", "NS"))
##multiple boxplot
ggplot(CD4.data.melt, aes(x = Gene, y = Expression, fill = tissue))+
  # geom_violin(trim = TRUE,alpha = 0.6)+
  
  # stat_boxplot(geom = "errorbar", width = 1)+
  geom_boxplot(outlier.shape = NA)+
  # geom_errorbar()+
  theme_bw() +
  ylab("Expression")+
  scale_fill_manual(values = c( "#377EB8","#E41A1C" ))+
  annotate("text",x = 1:length(CD4.genes), y = as.numeric(apply(CD4.data,1,max)[1:length(CD4.genes)]), label = p.star,size = 8,colour = "red")+
  theme(plot.title = element_text(hjust=0.5,vjust = 0.5, margin = margin(l=100,r=50,t=10,b=10),face = "bold", colour = "black"),
        axis.text.x=element_text(size=12,face = "bold",hjust=0.5),
        axis.text.y=element_text(size=12,face = "bold",hjust=0.5),
        axis.title.x = element_blank(),
        axis.title.y = element_text(size = 12, face = "bold"),
        legend.position = "right")
```

```{r, fig.height = 4, fig.width =11}
##############-------------------Figure 3: TCR/BCR repertoire similarity in one plot with both background----------------------##########
###########################################################################################################################
###prepare BCR data
type <- "BCR"
load(file = paste("~/Documents/Description_study/data/MIA_TN_",type,"_share_mat.Rdata",sep = ""))
load(file = paste("~/Documents/Description_study/data/MIA_paired_TN_",type,"_Jacard.Rdata",sep = ""))
bcr.shaired.pair <- shared.pair %>% mutate(Type = type, jacard = jacard) #%>% mutate(cluster = row_split[patient,"Cluster"])
backgroud.share <- uniqdf(share.mat,bcr.shaired.pair[,colnames(share.mat)])
bcr.backgroud.share.new <- backgroud.share %>% group_by(tumor) %>% 
  dplyr::summarise(jacard = signif(mean(jacard),2)) %>%
  mutate(patient = sapply(tumor,function(x)paste("Pat",patient.info[which(patient.info$tumor == x),"patientID"],sep = ""))) #%>% mutate(cluster = row_split[patient,"Cluster"])
###prepare TCR data
type <- "TCR"
load(file = paste("~/Documents/Description_study/data/MIA_TN_",type,"_share_mat.Rdata",sep = ""))
load(file = paste("~/Documents/Description_study/data/MIA_paired_TN_",type,"_Jacard.Rdata",sep = ""))
tcr.shaired.pair <- shared.pair %>% mutate(Type = type,jacard = -jacard) #%>% mutate(cluster = row_split[patient,"Cluster"])
backgroud.share <- uniqdf(share.mat,tcr.shaired.pair[,colnames(share.mat)])
tcr.backgroud.share.new <- backgroud.share %>% group_by(tumor) %>% 
  dplyr::summarise(jacard = signif(mean(jacard),2)) %>%
  mutate(patient = sapply(tumor,function(x)paste("Pat",patient.info[which(patient.info$tumor == x),"patientID"],sep = ""))) # %>% mutate(cluster = row_split[patient,"Cluster"])
###combine data and order patients
comb.shared.pari <- rbind.data.frame(bcr.shaired.pair,tcr.shaired.pair)  #  
total.jard <- comb.shared.pari %>% group_by(patient) %>% dplyr::summarise(overall.jard = sum(abs(jacard)))
comb.shared.pari$patient <- factor(comb.shared.pari$patient,
                                   levels = levels(reorder(total.jard$patient, as.numeric(total.jard$overall.jard))))
MIA.jacard.pat.order <- levels(reorder(total.jard$patient, as.numeric(total.jard$overall.jard)))
###two histograms plot
brks <- seq(-0.15, 0.15, 0.03)
lbls <- c(seq(0.15,0,-0.03), seq(0.03,0.15,0.03))
mainplt <- ggplot(comb.shared.pari,aes(x= patient, y = jacard, fill = Type))+
  geom_bar(stat = "identity",alpha = 0.8)+
  theme_bw()+
  annotate("text", x = 10, y = 0.12, label = "BCR Repertoire", size = 6, color = brewer.pal(n = 3, name = "Dark2")[2])+
  annotate("text", x = 10, y = -0.12, label = "TCR Repertoire", size = 6, color = brewer.pal(n = 3, name = "Dark2")[1])+
  scale_y_continuous(breaks = brks, labels = lbls)+
  scale_fill_manual(values = brewer.pal(n = 3, "Dark2")[2:1])+
  labs(x = NULL, y = "Jacard index")+
  theme(legend.position = "none",
        axis.text.x=element_blank(),
        axis.text.y = element_text(size = 12,face = "bold"),
        axis.title.x = element_blank(),
        axis.title.y = element_text(size = 12, face = "bold"),
        axis.ticks.x.bottom = element_blank()
  )

###backgroud plot for tcr and bcr
##for bcr
bcr.backgroud.share.new$patient <- factor(bcr.backgroud.share.new$patient, levels = levels(reorder(total.jard$patient, as.numeric(total.jard$overall.jard))))
backplt.bcr <- ggplot(bcr.backgroud.share.new,aes(x= patient, y = jacard))+
  geom_bar(stat = "identity",alpha = 0.5, fill = brewer.pal(n = 3, name = "Dark2")[2])+
  theme_bw()+
  labs(x = NULL, y = "")+
  scale_y_continuous(breaks = seq(0,0.03,0.01))+
  theme(legend.position = "none",
        axis.text.x=element_blank(),
        axis.text.y = element_text(size = 12,face = "bold"),
        axis.title.x = element_blank(),
        axis.title.y = element_text(size = 12, face = "bold"),
        axis.ticks.x.bottom = element_blank()
  )
##for tcr
tcr.backgroud.share.new$patient <- factor(tcr.backgroud.share.new$patient, levels = levels(reorder(total.jard$patient, as.numeric(total.jard$overall.jard))))
backplt.tcr <- ggplot(tcr.backgroud.share.new,aes(x= patient, y = jacard))+
  geom_bar(stat = "identity",alpha = 0.5, fill = brewer.pal(n = 3, name = "Dark2")[1])+
  theme_bw()+
  labs(x = NULL, y = "")+
  scale_y_continuous(breaks = seq(0,0.006,0.003))+
  theme(legend.position = "none",
        axis.text.x=element_blank(),
        axis.text.y = element_text(size = 12,face = "bold"),
        axis.title.x = element_blank(),
        axis.title.y = element_text(size = 12, face = "bold"),
        axis.ticks.x.bottom = element_blank()
  )
###combined plots
ggarrange(backplt.bcr, mainplt, backplt.tcr, ncol=1, nrow=3, widths = c(4.4,4.1,4),heights = c(1,4,1))

```


```{r,fig.height=4, fig.width=4.5}
##################-------------Figure 3: repertoire comparison on patient level for MIA----------##########################
###########################################################################################################################
TNhexbin <- function(type, cutmin, annopos){
  # load(file = paste("~/Documents/Description_study/data/MIA_pat_complete_clone_",type,".Rdata",sep = ""))
  # pat.clone.all <- do.call("rbind",pat.clone)
  # pat.clone.all <- get(load(file = paste("~/Documents/Description_study/data/MIA_pat_complete_all_clone_",type,".Rdata",sep = "")))
  pat.clone.all <- read.table(file = paste("~/Documents/Description_study/data/MIA_pat_complete_all_clone_",type,".txt",sep = ""),
                              sep = "\t", header = TRUE)
  pat.cluster.clones <- dcast(pat.clone.all,clonetype+patient+clone.size~Tissue, value.var = "total.freq.normalize")
  pat.cluster.clones[is.na(pat.cluster.clones)] <- cutmin    ####given a very small value for log transformation
  pat.cluster.clones <- dplyr::arrange(pat.cluster.clones,desc(clone.size))
  ##scatter plot
  cluster.t <- subset(pat.cluster.clones, Normal == cutmin & Tumor != cutmin)   ####highly for tumors
  cluster.n <- subset(pat.cluster.clones, Tumor == cutmin & Normal != cutmin)   ####highly for normals
  cluster.share <- subset(pat.cluster.clones,Normal != cutmin & Tumor != cutmin)
  t.perc <- paste(signif(dim(cluster.t)[1]/dim(pat.cluster.clones)[1],3)*100,"%",sep = "")
  n.perc <- paste(signif(dim(cluster.n)[1]/dim(pat.cluster.clones)[1],3)*100,"%",sep = "")
  share.perc <- paste(signif(dim(cluster.share)[1]/dim(pat.cluster.clones)[1],3)*100,"%",sep = "")
  ###hexbin and density plot
  p <- ggplot(pat.cluster.clones, aes(y = Tumor, x = Normal))+
    #stat_binhex(colour="white",na.rm=TRUE)+
    scale_fill_gradientn(colours = rev(RColorBrewer::brewer.pal(11, "Spectral")),name = "count",trans = "log2")+
    stat_binhex(bins = 64)+
    geom_density2d(colour = "black")+
    #scale_fill_gradientn(colours = brewer.pal(6, "OrRd"), trans = "log10",name = "count")+
    #ylim(0,0.1)+
    #geom_abline(intercept = -2, slope = 0.5)+
    scale_x_continuous(trans = log10_trans(),
                       breaks = trans_breaks("log10", function(x) 10^x),
                       labels = trans_format("log10", math_format(10^.x)))+
    scale_y_continuous(trans = log10_trans(),
                       breaks = trans_breaks("log10", function(x) 10^x),
                       labels = trans_format("log10", math_format(10^.x)))+
    annotate("label",x = annopos, y = 0.1, label = t.perc)+
    annotate("label",x = 0.01, y = annopos, label = n.perc)+
    annotate("label",x = 0.01, y = 0.1, label = share.perc)+
    #ggtitle(paste(Tissue," enriched motifs mapped to CDR3s",sep = ""))+
    xlab("Clone frequency in Normal")+
    ylab("Clone frequency in Tumor")+
    ggtitle(toupper(type))+
    theme_bw()+
    theme(axis.ticks.y = element_blank(),
          axis.ticks.x = element_blank(),
          axis.text.x = element_text(hjust = 1,size=12,face = "bold"),
          axis.text.y = element_text(size=12,face = "bold"),
          axis.title.x = element_text(hjust = 1,size=12,face = "bold"),
          axis.title.y = element_text(size=12,face = "bold"),
          plot.title = element_text(hjust=0.5,vjust = 0.5, margin = margin(l=50,r=50,t=10,b=10),face = "bold", colour = "black"))
  print(p)
}
TNhexbin(type = "bcr", cutmin = 0.000001, annopos = 0.000003)
TNhexbin(type = "tcr", cutmin = 0.000001, annopos = 0.000003)
```


```{r, fig.height = 4, fig.width = 12}
##############-------------------Figure 3: show T-T and N-N similarity for TCR and BCRs----------------------##########
#######################################################################################################################
load(file = "~/Documents/Description_study/data/MIA_jacard_pat_order.Rdata")
## similarity comparison
types <- c("TCR","BCR")
comps <- c("T_T", "N_N")
alljacard <- lapply(types, function(t){
  allcomps <- lapply(comps, function(c){
    tmp <- get(load(paste("~/Documents/Description_study/data/MIA_",c,"_",t,"_jacard_similarity.Rdata",sep = "")))
    tmp.tc <- tmp %>% mutate(type = t, compare = c)
  })
  allcomps.mat <- do.call("rbind", allcomps)  
  return(allcomps.mat)
})
alljacard.mat <- do.call("rbind", alljacard) #%>% mutate(jacard_axis = ifelse(type == "TCR", -jacard.mean, jacard.mean))
alljacard.mat$patient <- factor(alljacard.mat$patient,levels = MIA.jacard.pat.order)
###generate two sub figures
alljacard.mat.sub1 <- subset(alljacard.mat, type == "BCR")
g1 <- ggplot(alljacard.mat.sub1,aes(x= patient, y = jacard.mean, group = compare))+
  geom_line(aes(linetype = compare), size = 1, show.legend = TRUE,color = brewer.pal(n = 3, name = "Dark2")[2], alpha = 0.5) +
  geom_point(size = 2, show.legend = TRUE,color = brewer.pal(n = 3, name = "Dark2")[2]) +
  theme_bw()+
  annotate("text",x = 6, y = 0.0001, label = "BCR Repertoire", color = brewer.pal(n = 3, name = "Dark2")[2], size = 6)+
  scale_y_continuous(trans = log10_trans(),
                     breaks = trans_breaks("log10", function(x) 10^x),
                     labels = trans_format("log10", math_format(10^.x)))+
  labs(x = "patient", y = "Jacard index")+
  theme(legend.position = "right",
        axis.text.x=element_blank(),
        axis.text.y = element_text(size = 12,face = "bold"),
        axis.title.x = element_blank(),
        axis.ticks.x = element_blank(),
        axis.title.y = element_text(size = 12, face = "bold"))
alljacard.mat.sub2 <- subset(alljacard.mat, type == "TCR")
g2 <- ggplot(alljacard.mat.sub2,aes(x= patient, y = jacard.mean, group = compare))+
  geom_line(aes(linetype = compare), size = 1, show.legend = TRUE,color = brewer.pal(n = 3, name = "Dark2")[1], alpha = 0.5) +
  geom_point(size = 2, show.legend = TRUE,color = brewer.pal(n = 3, name = "Dark2")[1]) +
  # scale_y_reverse()+
  theme_bw()+
  annotate("text",x = 6, y = 0.01, label = "TCR Repertoire", color = brewer.pal(n = 3, name = "Dark2")[1], size = 6)+
  scale_y_continuous(trans = log10_trans(),
                     breaks = trans_breaks("log10", function(x) 10^x),
                     labels = trans_format("log10", math_format(10^.x)))+
  # scale_y_reverse()+
  labs(x = "patient", y = "Jacard index")+
  theme(legend.position = "right",
        axis.text.x=element_blank(),
        axis.text.y = element_text(size = 12,face = "bold"),
        axis.title.x = element_text(size = 12,face = "bold"),
        axis.ticks.x = element_blank(),
        axis.title.y = element_text(size = 12, face = "bold"))
## combine plots
ggarrange(g1,g2,ncol = 1, nrow = 2, common.legend = TRUE)
```


```{r, fig.height = 4, fig.width = 5.5}
##############-----------Figure 3: check whether the mutation load correlates TN repertoire similarity--------------##########
###########################################################################################################################
TBCR_Jacard <- function(type){
  load(file = paste("~/Documents/Description_study/data/MIA_paired_TN_",type,"_Jacard.Rdata",sep = ""))
  ##extract data
  #shared.pair <- rbind(tcr.shaired.pair, bcr.shaired.pair)
  jacard <- subset(shared.pair, select = c(normal, tumor, patient, jacard)) 
  ##mutation load
  oncomatrix.output <- read.table("~/Documents/Description_study/vcf2maf_data/early_lung_maftools_removeOutlier_oncomatrix.txt",sep = "\t",header = TRUE,fill = TRUE)
  #oncomatrix.output <- oncomatrix.output %>% dplyr::filter(!grepl("^IG",rownames(oncomatrix.output)))
  colnames(oncomatrix.output) <-  paste("Pat",pat.info[match(colnames(oncomatrix.output), gsub("-",".",pat.info$value)),'patientID'],sep = "")
  mut.load <- apply(oncomatrix.output,2, function(x) length(intersect(which(x != ""),which(!grepl("Frame_Shift",x)))))
  ss <- intersect(colnames(oncomatrix.output), jacard$patient)
  ##group comparison
  ct <- 0.5
  #lab <- c("0-25%","25%-50%","50%-75%","75%-100%")
  lab <- c("Low_ML","High_ML")
  ML <- mut.load[ss]
  names(ML) <- as.character(cut(ML,
                                breaks = quantile(ML, probs = seq(0, 1, ct)),
                                include.lowest = TRUE,
                                labels = lab))  ###four groups
  
  df.cl.ord <- cbind.data.frame(idex  = unlist(jacard[match(ss, jacard$patient),"jacard"]), 
                                isc = ML,
                                group = names(ML),
                                type = type)
  return(df.cl.ord)
}
###combine TCR and BCR jacard index together to see the corrlation between jacard and mutation load
com.df.cl.ord <- rbind.data.frame(TBCR_Jacard("TCR"), TBCR_Jacard("BCR"))
ggboxplot(com.df.cl.ord, x = "type", y = "idex",color = "group", palette = c("#377EB8","#E41A1C"),add = "jitter")+
  stat_compare_means(aes(group = group),method = "wilcox.test", label = "p.signif",
                     label.y = max(com.df.cl.ord$idex)-0.01,size = 5,
                     symnum.args = list(cutpoints = c(0, 0.001, 0.01, 0.05, 1), symbols = c( "***", "**", "*", "ns")))+
  xlab("Mutation load")+
  ylab("Jacard_Index")+
  ggtitle("")+
  theme_bw()+
  theme(plot.title = element_text(hjust=0.5,vjust = 0.5, margin = margin(l=100,r=50,t=10,b=10),face = "bold", colour = "black"),
        axis.text.x=element_text(angle=70,size=12,face = "bold",hjust=1),
        axis.text.y=element_text(size=12,face = "bold",hjust=1),
        axis.title.x = element_text(size = 12, face = "bold"),
        axis.title.y = element_text(size = 12, face = "bold"))

```


```{r, fig.width=15, fig.height=7}
#################------------Figure 4: rank patients from immune low to immune high (latest)-------------##########################
#############################################################################################################################
### TPM expression normalized by normal samples
pat.expr <- t.TPM[,t.sample]-t.TPM[,n.sample]
colnames(pat.expr) <- paste("Pat",pat.info[match(colnames(pat.expr), gsub("-",".",pat.info$value)),'patientID'],sep = "")
###loading module score
load(file = "~/Documents/Description_study/data/CRI_MIA_immune_module_score.Rdata")
score.aver <- apply(scores, 2, mean)
sc.rnk <- rank(score.aver)
score.rank <- scores[,names(sc.rnk)[match(c(1:length(score.aver)), sc.rnk)]] ##rank patient from low to high based on the averaged immune score
pat.ord <- names(score.aver)
names(pat.ord) <- ifelse(score.aver >= median(score.aver), "Immune_High","Immune_Low")
merge.moduls <- structure(names = rownames(scores),c("Lymphocyte","Macrophage","TGF_b","IFN_g","Wound_healing"))
col_split <- data.frame(Cluster = merge.moduls)
rownames(col_split) <- rownames(scores)
rownames(score.rank) <- as.character(unlist(col_split[rownames(score.rank),"Cluster"]))
###tumor purity
MIA.epic <- read.table(file = "~/Documents/Description_study/data/MIA_TPM_all_EPIC.txt",sep = "\t",header = TRUE,skip = 5, row.names = 1,check.names = FALSE)
purity <- MIA.epic$otherCells
names(purity) <- rownames(MIA.epic)
purity.tumor <- purity[paste("FZ.",tumor, sep = "")]
names(purity.tumor) <- paste("Pat",pat.info[match(names(purity.tumor), gsub("-",".",pat.info$value)),'patientID'],sep = "")
purity.normal <- purity[paste("FZ.",normal, sep = "")]
names(purity.normal) <- paste("Pat",pat.info[match(names(purity.normal), gsub("-",".",pat.info$value)),'patientID'],sep = "")
###rank by overall immune score
ord <- colnames(score.rank)   
####side bar of immune marker expression or infiltration
g00 <- score.aver[ord]
g03 <- purity.tumor[ord]
g1 <- unlist(pat.expr["CD40",ord])
g2 <- unlist(pat.expr["CD40LG",ord])
g3 <- unlist(apply(pat.expr[grep("^HLA-D",rownames(pat.expr),value = TRUE),ord],2,mean))
g4 <- unlist(apply(pat.expr[grep("^MUC",rownames(pat.expr),value = TRUE),ord],2,mean))
g6 <- unlist(apply(pat.expr[grep("^IGHG[1-4]",rownames(pat.expr),value = TRUE),ord],2,mean)) # IgG[ord] #
g7 <- unlist(apply(pat.expr[grep("^IGHA",rownames(pat.expr),value = TRUE),ord],2,mean)) # IgA[ord]
###overall immune score
ha <- columnAnnotation(Immune_Average = anno_lines(score.aver[colnames(score.rank)], add_points = TRUE,smooth = TRUE),
                       show_annotation_name = TRUE, annotation_name_side = "right", annotation_name_rot = 0)
###annotation bar
ha3 = HeatmapAnnotation(
  Immune_score = anno_lines(score.aver[colnames(score.rank)], add_points = TRUE,smooth = TRUE, height = unit(1.5, "cm")),
  Tumor_purity = g03,
  Mucin = g4,
  MHCII = g3,
  CD40 = g1,
  CD40LG = g2,
  IgG = g6,
  IgA = g7,
  col = list(
    Tumor_purity = colorRamp2(c(min(g03), (min(g03)+max(g03))/2, max(g03)), c("gray", "white", "black")),
    Mucin = colorRamp2(c(min(g4), 0, max(g4)), c("#3288BD", "white", "#D53E4F")),  #blue, red
    MHCII = colorRamp2(c(min(g3), 0, max(g3)), c("#3288BD", "white", "#D53E4F")),
    CD40 = colorRamp2(c(min(g1), 0, max(g1)), c("#66C2A5", "white", "#FC8D62")),
    CD40LG =  colorRamp2(c(min(g2), 0, max(g2)), c("#66C2A5", "white", "#FC8D62")),
    IgG = colorRamp2(c(min(g6), (min(g6)+max(g6))/2, max(g6)), c("white", "#A6CEE3", "#1F78B4" )), #white  #1F78B4
    IgA = colorRamp2(c(min(g7), (min(g7)+max(g7))/2, max(g7)), c("white", "#A6CEE3", "#1F78B4" ))
  ),
  na_col = "white", border = TRUE,
  show_legend = c(TRUE, TRUE,FALSE, TRUE,FALSE,TRUE,FALSE),  #
  show_annotation_name = TRUE, 
  annotation_name_side = "left",annotation_name_rot = 0,
  annotation_legend_param = list(
    Tumor_purity = list(title = "tumor_purity"),
    Mucin = list(title = "Mucin_MHC_Expression"),
    CD40 = list(title = "CD40_CD40L_Expression"),
    IgG = list(title = "IgG_IgA_Expression")
    )
)
#####heatmap for showing immune related signatures
ht_list = Heatmap(score.rank[,ord],  name = "Immune score",
                  cluster_rows = FALSE,
                  cluster_columns = FALSE,
                  show_column_names = FALSE,
                  show_column_dend = FALSE,
                  show_row_names = TRUE,
                  show_row_dend = FALSE,
                  row_names_side = "left", 
                  row_names_gp = gpar(fontsize = 12),
                  row_split = col_split,
                  bottom_annotation = ha3,
                  height = 1,
                  heatmap_height = 10
) 
draw(ht_list,annotation_legend_side = "right", heatmap_legend_side = "right",merge_legends = TRUE)

```


```{r,fig.height=4, fig.width=4}
###############--------------Figure 4: correlation between Mucin level and MHCII level----------------###########
############################################################################################
dat <- data.frame(F1 = unlist(apply(pat.expr[grep("^HLA-D",rownames(pat.expr),value = TRUE),pat.ord],2,mean)),
                  F2 = unlist(apply(pat.expr[grep("^MUC",rownames(pat.expr),value = TRUE),pat.ord],2,mean)),
                  F3 = names(pat.ord))
cor.test <- cor.test(dat$F1,dat$F2,method = "spearman")
cor <- signif(cor.test$estimate,2)
p <- cut(cor.test$p.value,breaks = c(0, 0.001, 0.01, 0.05,1), labels = c("***","**","*"," "))
ggplot(dat,aes(x = F2, y = F1))+ #, color = F3
  geom_point()+
  geom_smooth(aes(x=F2, y=F1), method = lm, color = "blue")+
  theme_bw()+
  xlab("Mucin_Level")+
  ylab("MHCII_Level")+
  ggtitle(paste(paste(paste("Correlation = ", cor, sep = ""), p, sep = "  ")))+
  theme(legend.position = "top",
        plot.title = element_text(hjust=0.5,vjust = 0.5, margin = margin(l=100,r=50,t=10,b=10),face = "bold", colour = "black"),
        axis.text.x=element_text(angle=70,size=12,face = "bold",hjust=1),
        axis.text.y=element_text(size=12,face = "bold",hjust=1),
        axis.title.x = element_text(size=12,face = "bold"),
        axis.title.y = element_text(size = 12, face = "bold")
  ) 
```


```{r, fig.height=4, fig.width=4}
###############--------------Figure 4: correlation between CD40 level and CD40LG level----------------###########
############################################################################################
dat <- data.frame(F1 = unlist(pat.expr["CD40",]),
                  F2 = unlist(pat.expr["CD40LG",]),
                  F3 = names(pat.ord))
cor.test <- cor.test(dat$F1,dat$F2,method = "spearman")
cor <- signif(cor.test$estimate,2)
p <- cut(cor.test$p.value,breaks = c(0, 0.001, 0.01, 0.05,1), labels = c("***","**","*"," "))
ggplot(dat,aes(x = F2, y = F1))+ #, color = F3
  geom_point()+
  geom_smooth(aes(x=F2, y=F1), method = lm, color = "blue")+
  theme_bw()+
  xlab("CD40LG_Level")+
  ylab("CD40_Level")+
  ggtitle(paste(paste(paste("Correlation = ", cor, sep = ""), p, sep = "  ")))+
  theme(legend.position = "top",
        plot.title = element_text(hjust=0.5,vjust = 0.5, margin = margin(l=100,r=50,t=10,b=10),face = "bold", colour = "black"),
        axis.text.x=element_text(angle=70,size=12,face = "bold",hjust=1),
        axis.text.y=element_text(size=12,face = "bold",hjust=1),
        axis.title.x = element_text(size=12,face = "bold"),
        axis.title.y = element_text(size = 12, face = "bold")
  ) 
```


```{r,fig.height=4, fig.width=4}
###############--------Figure 4: correlation between Mucin level and CD40L-CD40 level--------###########
############################################################################################
dat <- data.frame(F1 = unlist(apply(pat.expr[c("CD40","CD40LG"),pat.ord],2,mean)),
                  F2 = unlist(apply(pat.expr[grep("^MUC",rownames(pat.expr),value = TRUE),pat.ord],2,mean)),
                  F3 = names(pat.ord))
cor.test <- cor.test(dat$F1,dat$F2,method = "spearman")
cor <- signif(cor.test$estimate,2)
p <- cut(cor.test$p.value,breaks = c(0, 0.001, 0.01, 0.05,1), labels = c("***","**","*"," "))
ggplot(dat,aes(x = F2, y = F1))+ #, color = F3
  geom_point()+
  geom_smooth(aes(x=F2, y=F1), method = lm, color = "blue")+
  theme_bw()+
  xlab("Mucin_Level")+
  ylab("CD40-CD40L_Level")+
  ggtitle(paste(paste(paste("Correlation = ", cor, sep = ""), p, sep = "  ")))+
  theme(legend.position = "top",
        plot.title = element_text(hjust=0.5,vjust = 0.5, margin = margin(l=100,r=50,t=10,b=10),face = "bold", colour = "black"),
        axis.text.x=element_text(angle=70,size=12,face = "bold",hjust=1),
        axis.text.y=element_text(size=12,face = "bold",hjust=1),
        axis.title.x = element_text(size=12,face = "bold"),
        axis.title.y = element_text(size = 12, face = "bold")
  ) 
```


```{r, fig.height=4, fig.width=4}
#############--------------------Figure 4: scatterplot of mucin expression and Ig abundance-------------------#################
###########################################################################################################################
##combined data
marker <- "Mucin"
df.combine <- rbind.data.frame(cbind.data.frame(G1 = g4, G2 = g6, Group = "IgG abundance"),
                               cbind.data.frame(G1 = g4, G2 = g7, Group = "IgA abundance")
                               )
##p value and correlation
spe <- split(df.combine, df.combine$Group,drop=FALSE) 
Pcor <- vector("list", length = length(spe))
Pcor <- lapply(1:length(spe), function(x){
  df <- spe[[x]]
  test <- cor.test(unlist(df$G1),unlist(df$G2), method = "spearman")
  res <- c(Group = as.character(unique(df$Group)), 
           pval = as.character(cut(test$p.value,breaks = c(0, 0.001, 0.01, 0.05,1), labels = c("***","**","*"," "))), 
           corrlation = signif(test$estimate,2),
           Gene = marker)
  return(res)
})
Pcor.res <- do.call("rbind",Pcor)
##scatter plot
ggplot(df.combine,aes(x = G1, y = G2)) +
  geom_point(aes(colour = Group))+
  geom_smooth(aes(group=Group,color = Group),method=lm) +
  theme_bw() +
  ylab("Ig_Level")+
  scale_color_manual(values = RColorBrewer::brewer.pal(n = 3, name = "Dark2"))+
  labs(x = paste(marker,"_Level"),sep = "")+
  annotate("text",x = max(df.combine$G1)*0.10, y = max(df.combine$G2)*0.95,label = paste(Pcor.res[1,1], ":",Pcor.res[1,3],Pcor.res[1,2],sep = " "),color = brewer.pal(n = 3, name = "Dark2")[1])+
  annotate("text",x = max(df.combine$G1)*0.08, y = max(df.combine$G2)*0.80,label = paste(Pcor.res[2,1], ":",Pcor.res[2,3],Pcor.res[2,2],sep = " "),color = brewer.pal(n = 3, name = "Dark2")[2])+
  # annotate("text",x = max(df.combine$G1)*0.12, y = max(df.combine$G2)*0.65,label = paste(Pcor.res[3,1], ":",Pcor.res[3,3],Pcor.res[3,2],sep = " "),color = brewer.pal(n = 3, name = "Dark2")[3])+
  theme(plot.title = element_text(hjust=0.5,vjust = 0.5, margin = margin(l=100,r=50,t=10,b=10),face = "bold", colour = "black"),
        axis.text.x=element_text(angle=70,size=12,face = "bold",hjust=1),
        axis.text.y=element_text(size=12,face = "bold",hjust=1),
        axis.title.x = element_text(size = 12, face = "bold"),
        axis.title.y = element_text(size = 12, face = "bold"),
        legend.position = "top")
```


```{r, fig.height=6, fig.width=14}
############################-------------Figure 5: gene expression for 6 ELS markers----------------########################
############################################################################################################################
###zscore-normalized TPM data
TPM.mean <- apply(t.TPM,1,mean)
TPM.sd <- apply(t.TPM,1,sd)
TPM.zscore <- apply(t.TPM, 2, function(x) {
  y <- as.numeric(x)
  zscore <- (y-TPM.mean)/TPM.sd
  return(zscore)
})
colnames(TPM.zscore) <- colnames(TPM)#paste("Pat",pat.info[match(paste("FZ-",tumor,sep = ""),pat.info$value),"patientID"],sep = "")
rownames(TPM.zscore) <- rownames(TPM)
###read in infiltration data
timer <- read.table(file = '~/Documents/BCR_Project_Xiaoqi/20171024-RNA-Seq data/score_matrix.csv',header = TRUE,sep = ",",row.names = 1)
timer <- t(timer)
###tumor purity
MIA.epic <- read.table(file = "~/Documents/Description_study/data/MIA_TPM_all_EPIC.txt",sep = "\t",header = TRUE,skip = 5, row.names = 1)
purity <- MIA.epic$otherCells
names(purity) <- rownames(MIA.epic)
###read in data
GC.genes <- c("CD19","MS4A1","CXCR5","CXCL13","CCR7","CCL19") 
cells <- c("B_cell","T_cell.CD4","T_cell.CD8")
GC.expr <- as.matrix(TPM.zscore[GC.genes,c(t.sample, n.sample)])
###compute correlation
CorrCellGene <- CorrelatinExprInfilCorrectPurity(cells,GC.genes,t.TPM[,t.sample],timer, purity, method = "+")
combined.corr <- CorrCellGene[which(grepl("\\/",CorrCellGene$Gene)),]
single.corr <- CorrCellGene[which(!grepl("\\/",CorrCellGene$Gene)),]
###prepare side correlation value between infiltration and separated gene expression 
correlation_col <- colorRamp2(c(-0.8,0,0.8), c("#377EB8", "white", "#E41A1C"))
ha.g2 = rowAnnotation(
  B_cell = single.corr[which(single.corr$cell == "B_cell"),"correlation"],
  CD4_T = single.corr[which(single.corr$cell == "T_cell.CD4"),"correlation"],
  CD8_T = single.corr[which(single.corr$cell == "T_cell.CD8"),"correlation"],
  col = list(B_cell = correlation_col,
             CD4_T = correlation_col,
             CD8_T = correlation_col),
  na_col = "white", #border = TRUE,
  show_legend = c(TRUE, FALSE, FALSE),
  show_annotation_name = TRUE, #c("CD19+MS4A1", "CXCR5+CXCL13", "CCR7+CCL19"),
  annotation_legend_param = list(B_cell = list(title = "Correlation")),
  gp = gpar(col = "white")
)
###top anntation for samples grouping
Tissue_col <- structure(names = c("Normal", "Tumor"), c("navy","firebrick3"))
ha.g3 <- columnAnnotation(
  Tissue = c(rep("Tumor",length(t.sample)),rep("Normal",length(n.sample))),
  col = list(Tissue = Tissue_col
  )
)
###grouping the genes
g.group <- c(rep("B cell Infiltration",2),rep("Follicle phenotype",2),rep("Lymphocyte homing signal",2))
ha.g4 = rowAnnotation(
  Group = g.group,
  col = list(Group = structure(names = unique(g.group), brewer.pal(n = 3, name = "Set3")))
)
###prepare main heatmap
ht_list = Heatmap(GC.expr, name = "Expression",
                  # col = col_fun,
                  clustering_distance_columns = "euclidean",
                  cluster_rows = FALSE,
                  show_column_dend = TRUE,
                  # show_row_dend = FALSE,# rect_gp = gpar(col= "white"), 
                  show_column_names = FALSE,#row_split = g.group,
                  row_names_side = "left", 
                  row_names_gp = gpar(fontsize = 12),
                  left_annotation = ha.g4, 
                  right_annotation = ha.g2,
                  top_annotation = ha.g3)
###draw together
draw(ht_list, padding = unit(c(20, 2, 2, 2), "mm"), 
     merge_legends = TRUE, heatmap_legend_side = "bottom", 
     annotation_legend_side = "bottom") 
```


```{r}
##########--------------Figure 6: check suppressice immune cells by Xcell----------------###########
#################################################################################
suppres.cells <- c("Macrophages.M2", "Neutrophils", "Monocytes", "Fibroblasts", "Tregs")
n = 5
xsize = 10
##for xcell
xcell <- read.table(file = "~/Documents/Description_study/data/xCell_lung_TPM_all_xCell_1245030118.txt",header = TRUE,sep = "\t",row.names = 1)
xcell <- t(xcell)[,c(1:64)]
Infil.est.xcell <- data.frame(sampleID = rownames(xcell),xcell,row.names = NULL)
Infil.est.xcell <- subset(Infil.est.xcell, !(sampleID %in% c("FZ.25","FZ.26",'FZ.31','FZ.32'))) %>% subset(select = c("sampleID", suppres.cells))
infil.melt.xcell <- melt(Infil.est.xcell, id.vars = "sampleID")
colnames(infil.melt.xcell) <- c("ID","Immune.cell","Estimation")
infil.melt.xcell$Group <- c(rep(pat.diagnosis,n))
##calculate significance 
dat1 <- Infil.est.xcell
pat.new <- data.frame(row.names = Infil.est.xcell$sampleID, pat.diagnosis)
dat1$Group <- pat.new$pat.diagnosis
all.p <- NULL
for(i in colnames(dat1)[2:(n+1)]){
  subdat <- subset(dat1,select = c(i,"Group"))
  colnames(subdat) <- c("cell","group")
  P <- wilcox.test(cell~group,data=subdat,paired = pair)$p.value
  all.p <- c(all.p,P)
}
p.star <- cut(all.p,breaks = c(-Inf, 0.001, 0.01, 0.05, Inf), label=c("***", "**", "*", "ns"))
##ggplot
ggplot(infil.melt.xcell,aes(x=Immune.cell, y=Estimation,fill=Group))+
  geom_boxplot(outlier.size = 0.5)+
  ggtitle("")+ 
  ylab("Immune cell Abundance")+
  scale_fill_manual(values = c("#377EB8","#E41A1C"))+
  annotate("text",x = 1:n, y = as.numeric(apply(Infil.est.xcell[2:(n+1)],2,max))+0.1, label = p.star, color = "red",size = 7)+
  theme_bw()+
  theme(plot.title = element_text(hjust=0.5,vjust = 0.5, margin = margin(l=100,r=50,t=10,b=10),face = "bold", colour = "black"),
        axis.text.x=element_text(angle=30,size=xsize,face = "bold",hjust=1),
        axis.text.y = element_text(size = 12,face = "bold"),
        axis.title.x = element_blank(),
        axis.title.y = element_text(size = 12,face = "bold"),
        legend.text = element_text(size = 12),
        legend.position='bottom')
```


```{r,fig.height = 4.5, fig.width = 4.5}
###########----------Figure 6: scatterplot of CD8 T cell infiltration, FOXP3/BCL6, IL2RA/IL7R,PDCD1/ICOS-----------##########
#############################################################################################################
gl <- list(c("BCL6","FOXP3"),c("ICOS","PDCD1"),c("IL7R","IL2RA"))
c <- "T_cell.CD8"
ge.rank.zscore <- TPM.zscore[,t.sample]
##combine data
CorrGes <- lapply(gl, function(g){
  ge.select <- apply(t.TPM[,t.sample],2,function(x)match(x,sort(x))) #/dim(ge.rank.zscore)[2]  ###rank of genes
  rownames(ge.select) <- rownames(t.TPM)
  ge.mul <- t.TPM[g,t.sample]  #*ge.select[g,]
  ge.com <- unlist(log2(ge.mul[2,]/ge.mul[1,]+1))
  ge.com <- (ge.com - min(ge.com))/(max(ge.com) - min(ge.com))
  #ge.select <- apply(ge.rank.zscore[g,],1,function(x)match(x,sort(x))) #/dim(ge.rank.zscore)[2]  ###rank of genes
  #ge.select <- apply(ge.select, 1, function(x) log2(x[[2]]/x[[1]]))  
  name <- paste(rev(g),collapse = "/")
  names(ge.com) <- t.sample
  #ge.rank <- sort(ge.select)
  dat <- cbind.data.frame(Expression = ge.com,
                          Infil = unlist(timer[c,names(ge.com)]),
                          Cell = c,
                          purity = purity[names(ge.com)],
                          Gene = name)
  return(dat)
})
CorrGes.dat <- do.call("rbind",CorrGes)
##computing correlation and p value
combined.corr <- CorrelatinExprInfilCorrectPurity(c,gl,t.TPM[,t.sample],timer,purity,method = "/")
##scatter plot
ggplot(CorrGes.dat,aes(x = Infil, y = Expression)) +
  geom_point(aes(colour = Gene))+
  geom_smooth(aes(group=Gene,color = Gene),method=lm) +
  theme_bw() +
  ylab("Markers Ratio")+
  scale_color_manual(values = brewer.pal(n = 3, name = "Dark2"),name = "")+
  labs(x = "CD8+ T Infiltration")+
  annotate("text",x = max(CorrGes.dat$Infil)*0.75, y = max(CorrGes.dat$Expression)*0.95,label = paste(combined.corr[1,1], ":",combined.corr[1,2],combined.corr[1,4],sep = " "),color = brewer.pal(n = 3, name = "Dark2")[1])+
  annotate("text",x = max(CorrGes.dat$Infil)*0.75, y = max(CorrGes.dat$Expression)*0.75,label = paste(combined.corr[2,1], ":",combined.corr[2,2],combined.corr[2,4],sep = " "),color = brewer.pal(n = 3, name = "Dark2")[2])+
  annotate("text",x = max(CorrGes.dat$Infil)*0.75, y = max(CorrGes.dat$Expression)*0.55,label = paste(combined.corr[3,1], ":",combined.corr[3,2],combined.corr[3,4],sep = " "),color = brewer.pal(n = 3, name = "Dark2")[3])+
  theme(plot.title = element_text(hjust=0.5,vjust = 0.5, margin = margin(l=100,r=50,t=10,b=10),face = "bold", colour = "black"),
        axis.text.x=element_text(angle=70,size=12,face = "bold",hjust=1),
        axis.text.y=element_text(size=12,face = "bold",hjust=1),
        axis.title.x = element_text(size = 12, face = "bold"),
        axis.title.y = element_text(size = 12, face = "bold"),
        legend.position = "top")
```


